model-optimizer/extensions/back/MarkNodesWithShapeValues.py

```diff
@@ -0,0 +1,88 @@
+
+    This is needed to enable conversion to FP16 even if values in ShapeOf subgraphs exceed max(float16)
+    or because of FP16 lower precession shape inference is incorrect on some nodes (e.g. if Interpolate in scales mode
+    accepts values from ShapeOf subgraph).
+
+    This transformation must be run after shape inference and after all transformations that insert/modify
+    Cast nodes in ShapeOf subgraphs therefore it's placed at the end of the back phase.
+    """
+    enabled = True
+    graph_condition = [lambda graph: graph.graph['cmd_params'].data_type == 'FP16']
+
+    def run_after(self):
+        from extensions.back.pass_separator import BackFinish
+        return [BackFinish]
+
+    def run_before(self):
+        return []
+
+    @staticmethod
+    def get_operations_with_shape_inputs():
+        return {
+            'Interpolate': [1, 2],  # sizes, scales inputs
+            'Reshape': [1],  # shape
+            'Broadcast': [1],  # target_shape
+            'ConvBackPropData ': [2],  # output_shape
+            'GroupConvolutionBackpropData ': [2],  # output_shape
+            'BatchToSpace': [1, 2, 3],  # block_shape, crops_begin, crops_end
+            'SpaceToBatch': [1, 2, 3],  # block_shape, pads_begin, pads_end
+            'StridedSlice': [1, 2, 3],  # begin, end, strides
+            'VariadicSplit': [2],  # split_lengths
+            'Tile': [1],  # repeats input
+            'TopK': [1],  # K input
+            'Pad': [1, 2],  # pads_begin, pads_end
+            'Range': [0, 1, 2],  # start, stop, step inputs
+            'OneHot': [1],  # depth input
+        }
+
+    def find_and_replace_pattern(self, graph: Graph):
+        shape_input_ops_map = self.get_operations_with_shape_inputs()
+        nodes_with_shape_inputs = []
+        for op_type in shape_input_ops_map:
+            nodes_with_shape_inputs.extend(graph.get_op_nodes(type=op_type))
+
+        start_nodes = []
+        for node in nodes_with_shape_inputs:
+            start_ports = [x for x in shape_input_ops_map[node.type] if not node.in_port(x).disconnected()]
+            start_nodes.extend([node.in_port(port_idx).get_source().node for port_idx in start_ports])
+
+        condition = lambda node: node.soft_get('type') != 'ShapeOf'
+        nodes_with_shape_values = MarkSubGraphsWithCorrectLayout.bfs(start_nodes, set(), condition, forward=False)
+        for node in nodes_with_shape_values:
+            node['returns_shape_value'] = True
+            if node.type == 'Const':
+                if node.value.dtype == np.float32:
+                    node.out_node(0)['correct_data_type'] = True
+                elif node.value.dtype in [np.float16, np.float64]:
+                    log.debug('Const nodes {} with shape values have {} type'.format(node.soft_get('name', node.id),
+                                                                                     node.value.dtype))
```

Comment:
Why do we need this check and log?

Reply:
If float16 or float64 Const suddenly appear in shape flow it is an untypical and suspicious event indicating that either the original model had such Consts or in some MO transformation we inserted such Consts. I would prefer not to change by force data_type to **FP32** for the following reasons:
1. maybe model developer did that for a reason.
2. if we inserted in some transformations such Consts then it's better to revise out transformations and this log message will unhide such transformations.

Reply:
I am sure that nobody will find this message in HUGE debug log :) Also, we have FP16 models from ONNX, so there is a high chance that FP16 appear in a shape sub-graph for them. Anyway, resolving the comment